src/index.tsx

```diff
@@ -306,3 +322,114 @@ const getSystemTheme = (e?: MediaQueryList | MediaQueryListEvent) => {
   const systemTheme = isDark ? 'dark' : 'light'
   return systemTheme
 }
+
+
+let Cookies: {
+  get: (name: string) => string | null;
+}
+try {
+  if (isServer) Cookies = require('next/headers').cookies()
+} catch(e) { 
+  Cookies = { get: (_: string) => null }
+}
+
+// Properties for rendering <html> on the server in a way that will match client after hydration
+const getThemeHtmlProps = ({
+  attribute = 'data-theme',
+  cookieName = 'theme',
+  defaultTheme = 'light',
+  enableColorScheme = true,
+  value,
+}: ThemeProviderProps) => {
+  const props: HTMLProps<HTMLHtmlElement> = {}
+
+  const resolved = Cookies.get(cookieName) || defaultTheme
+  const name = value ? value[resolved] : resolved
+
+  if (attribute === 'class') {
+    if (name) props.className = name
+  } else {
+    (props as Record<string, string>)[attribute] = name
+  }
+
+  if (enableColorScheme) {
+    const fallback = colorSchemes.includes(defaultTheme) ? defaultTheme : null
+    const colorScheme = colorSchemes.includes(resolved) ? resolved : fallback
+    props.style = { colorScheme } as CSSProperties
+  }
+
+  return props
+}
+
+// Wraps an <html> element on the server to apply themes before reaching the client
+export const ServerThemeProvider: React.FC<ThemeProviderProps> = ({ children, ...props }) => {
+  if (!children || (children as ReactElement).type !== 'html') {
+    throw new Error('<ServerThemeProvider> must contain the <html> element.')
+  }
+  const child = children as ReactElement
+  const original = child.props
+  const resolved = getThemeHtmlProps(props)
+  if (original.className && resolved.className) {
+    resolved.className = `${original.className} ${resolved.className}`
+  }
+  if (original.style && resolved.style) {
+    let originalStyle = original.style
+    if (typeof originalStyle === 'string') {
+      originalStyle = Object.fromEntries(originalStyle.split(/;\s*/).map(x => x.split(/\s*:\s*/)))
+    }
+    resolved.style = {
+      ...resolved.style,
+      ...originalStyle,
+    }
+  }
+
+  let newKids: ReactNodeArray
+  if (!original.children) {
+    newKids = []
+  } else if (!Array.isArray(original.children)) {
+    newKids = [original.children]
+  } else {
+    newKids = [...original.children]
```

Comment:
I'm semi-sure you can just one line the elses as 

`else { newKids = [...original.children]`.

If it's one element it will be inserted in the array, otherwise it will be spread.

Reply:
The type of children is `ReactNode`, which can theoretically be a string or any other primitive, so to be cautious I want to make sure it doesn't just iterate through it and accidentally create an array of characters or something weird.